// girc/lib/plugins/master.js
function Master(bot, config) {
	this.bot = bot;
	this.client = bot.client;
	this.emitter = bot.emitter;
	this.config = config;

	this.masters = [];
	this.masterTimers = [];
	this.masterAuths = config.auths;
	this.masterTimeout = config.timeout*1000;
}

Master.prototype.init = function() {
	var self = this;

	this.emitter.on('pm', function(from, message) {
		if (self.masters.indexOf(from) !== -1) {
			self.emitter.emit('master', from, message);
			self.__reset(from);
		}
		else if (message == '!auth') {
			self.bot.whois(from, function(user) {
				if (typeof user['account'] !== 'undefined') {
					if (self.masterAuths.indexOf(user['account']) !== -1) {
						self.__add(from);
						self.emitter.emit('regmaster', from);
					}
				}
			});
		}
	});
};

Master.prototype.__add = function(nick) {
	if (this.masters.indexOf(nick) === -1) {
		this.masters.push(nick);
		this.__reset(nick);
	}
};

Master.prototype.__reset = function(nick) {
	var self = this;
	
	clearTimeout(this.masterTimers[nick]);
	this.masterTimers[nick] = setTimeout(function() {
		self.__del(nick);
	}, this.masterTimeout);
};

Master.prototype.__del = function(nick) {
	if (this.masters.indexOf(nick) !== -1) {
		this.masters.splice(this.masters.indexOf(nick), 1);
		this.emitter.emit('unregmaster', nick);
	}
};

module.exports = Master;
