// middleirc/lib/plugins/nickserv.js
function NickServ(bot, config) {
	this.bot = bot;
	this.client = bot.client;
	this.emitter = bot.emitter;
	this.config = config;
}

NickServ.prototype.init = function() {
	var self = this;
	
	this.emitter.on('notice', function(from, message) {
		if (from == 'NickServ') {
			if (message.match(/This nickname is registered/i)) {
				self.emitter.emit('identifying');
				self.identify();
			}
			else if (message.match(/Password accepted|You are now identified/i)) {
				self.emitter.emit('identified');
			}
			else if (message.match(/Your nick has been logged out|You have been logged out/i)) {
				self.emitter.emit('unidentified');
				self.identify();
			}
			else if (message.match(/Your nick isn't registered|is not a registered nickname/i)) {
				self.emitter.emit('wrongnick');
			}
		}
	});
};

NickServ.prototype.identify = function() {
	this.bot.say('NickServ', 'identify '+this.config.password);
};

NickServ.prototype.logout = function() {
	this.bot.say('NickServ', 'logout');
};

module.exports = NickServ;
